### DISCUSSION TITLE
Browser window opens up outside of application

### DISCUSSION THREAD

**Discussion by bgoldstein14** [cefsharp/CefSharp#4145] [General]
When testing with a 98.x version of CefSharp and using identical code to our working codebase using a 91.x version, the browser window did not open up inside the application (WPF window) as expected but rather opened up the URL in an actual browser outside the application. Is there something new that needs to be configured in order to have the expected behavior of having the URL open up within the application?

**Comment by amaitland** [comment]
> When testing with a 98.x version of CefSharp

Does the problem reproduce with the current supported version? Version 102 at time of writing. 

> the browser window did not open up inside the application (WPF window) as expected but rather opened up the URL in an actual browser outside the application. 

What do you mean by this?

**Reply by bgoldstein14** [reply]
I have not tried the newer version, but I suspect it will have the same behavior unless the issue I'm seeing was specifically addressed. What happens is, the WPF code initializes the Chromium browser and it opens up the URL in an instance of Chrome outside of the application when the code was initializing for the Chromium browser to open up on a WPF page inside our application.

**Reply by amaitland** [reply]
https://github.com/cefsharp/CefSharp.MinimalExample is provided for testing purposes and is kept up to date.

**Reply by bgoldstein14** [reply]
Alex, I'll try it out. However, it probably will work the same way that I did get the newer version to work when I set the URL in the control itself and not let it set itself dynamically when the WPF window initializes, i.e., "var browser = new ChromiumWebBrowser(args.Url);. That bit of code in 98.x will launch a browser using the desired URL outside of the entire application. 

I can check this out in the latest version when I have a chance to switch gears and get into that. The main reason I brought this up is in case this was some sort of known issue and/or new configuration that's necessary to prevent the behavior we're seeing in these later versions of CefSharp.

**Reply by amaitland** [reply]
> Alex, I'll try it out. However, it probably will work the same way that I did get the newer version to work when I set the URL in the control itself and not let it set itself dynamically when the WPF window initializes, i.e., "var browser = new ChromiumWebBrowser(args.Url);. That bit of code in 98.x will launch a browser using the desired URL outside of the entire application.

Please test the minimal example unmodified first, then modify the source to replicate your code. 

I'd be very surprised if the behaviour you are describing reproduces. 

Are you calling Process.Start from within say lifespanhandler?

> The main reason I brought this up is in case this was some sort of known issue and/or new configuration that's necessary to prevent the behavior we're seeing in these later versions of CefSharp.

You are the only one reporting this behaviour.

**Reply by amaitland** [reply]
Were you able to reproduce the problem with the `MimimalExample`?

`WPF` version is working as expected for me.

**Reply by bgoldstein14** [reply]
Alex, we're not doing anything with Process.Start. I also tried upgrading to the latest version and the same behavior occurred (the URL opened up the web page in my MS Edge browser instead of on the WPF page). If I simply set the address inside the control in the XAML, then the control works. The minimal example does not use MVVM as our solution requires and does not replicate our situation where the address is being set by the explicit initialization of the browser from viewmodel code. Once I have some time, I will try to replicate this situation in a smaller project that I can upload here. Thank you for being so responsive...that's much appreciated.

**Reply by amaitland** [reply]
> Once I have some time, I will try to replicate this situation in a smaller project that I can upload here. 

Please fork the minimal example and use that as a base rather than creating a new project. Push changes to GitHub. 

I won't review zip/compressed solutions as they are impractical to open on my android tablet.

**Reply by bgoldstein14** [reply]
Alex, a bit of good news on the upgrade front...

We finally had some time to do another review of the situation with the CefSharp component opening up the URL in the external browser (not the Chromium one in the app). I also did this with the latest stable version 105.

What we discovered is the since our link goes to a site which requires the user to be logged in via Microsoft single sign-on that CefSharp issued a redirect which then turned into a link clicked transaction type. We have a browser handler setup so that the link clicked transaction will open up the URL as an external browser transaction, i.e. "Process.Start(request.Url);" which is the exact behavior we're seeing and the initial reason for this thread. What's interesting is that for some reason it needs to iterate through the redirect/link-clicked transactions a couple times when it tries to traverse the SSO request chain.

As a solution which we started to play around with, we'll trap for the LinkedClicked transaction type, determine if there's a Referral URL and note that the actual URL contains something like "login.microsoftonline.com" and then let it continue along its normal browser opening functionality.

We'll be probably doing this upgrade relatively soon. The only other thing that we already know that will require a bit of extra research/work for us is how to get the VC++ redistributables out with our click-once deployment (via ADO) for this WPF application.

The interesting part in all of this is that our older version of CefSharp does not experience this behavior but something must have changed in one of the newer versions where it was somehow more sensitive to whatever SSO is doing.

**Reply by amaitland** [reply]
> We have a browser handler setup so that the link clicked transaction will open up the URL as an external browser transaction, i.e. "Process.Start(request.Url);" which is the exact behavior we're seeing and the initial reason for this thread

That's pretty much as I expected. The browser on it's own won't open a URL in an external browser. 

> versions where it was somehow more sensitive to whatever SSO is doing.

From memory there were some SSO changes recently, try searching https://bitbucket.org/chromiumembedded/cef/ if you want to know more.

> The only other thing that we already know that will require a bit of extra research/work for us is how to get the VC++ redistributables out with our click-once deployment (via ADO) for this WPF application.

Depending on your usage scenario you maybe able to bin deploy the runtime. See https://github.com/cefsharp/CefSharp/issues/3636#issuecomment-889580410
